### PR TITLE
feat: add fancy collapsible header logic

### PR DESCRIPTION
**Pull request by MounirDhahri**
### Description
This PR adds logic to support a fancy collapsible header. This is needed for the upcoming work on the `Activity` screen and potentially for other similar screens. Ideally, we can have a similar logic for all the screens 🤞

### How to use
Screenshot 2024-01-17 at 11 34 03

### Customization

Untitled-2023-04-27-1400

### Demo

https://github.com/artsy/eigen/assets/11945712/7a198180-9646-4073-bf2b-03bacd4d086a

### PR Checklist

- [x] I have tested my changes on **iOS** and **Android**.
- [x] I hid my changes behind a **[feature flag]**, or they don't need one.
- [x] I have included **screenshots** or **videos**, or I have not changed the UI.
- [x] I have added **tests**, or my changes don't require any.
- [x] I added an **[app state migration]**, or my changes do not require one.
- [x] I have documented any **follow-up work** that this PR will require, or it does not require any.
- [x] I have added a **changelog entry** below, or my changes do not require one.

### To the reviewers 👀

- [ ] I would like **at least one** of the reviewers to **run** this PR on the simulator or device.

<details><summary>Changelog updates</summary>

#### Cross-platform user-facing changes

- add fancy collapsible header logic - mounir

#### iOS user-facing changes

-

#### Android user-facing changes

-

#### Dev changes

-

</details>

Need help with something? Have a look at our [docs], or get in touch with us.

[app state migration]: ../blob/main/docs/adding_state_migrations.md
[feature flag]: ../blob/main/docs/developing_a_feature.md
[docs]: ../blob/main/docs/README.md

### CODE REVIEW

File: src/app/utils/useFancyHeader.tsx [hunk 1]
```diff
@@ -0,0 +1,227 @@
...
+const DYNAMIC_CONTENT_HEIGHT_MAX = 120
+const DYNAMIC_CONTENT_HEIGHT_MIN = 56
+const STICKY_CONTENT_HEIGHT = 50
+const DYNAMIC_CONTENT_TITLE_FONT_SIZE = parseInt(TEXT_VARIANTS["lg-display"].fontSize, 10)
+const NAVIGATION_TITLE_FONT_SIZE = parseInt(TEXT_VARIANTS["sm"].fontSize, 10)
+const DYNAMIC_CONTENT_TITLE_INCREASE_SIZE = 4
+
+const HORIZONTAL_PADDING = 20
+
+export const useFancyHeader = ({
```
Comment (MounirDhahri) [inline src/app/utils/useFancyHeader.tsx:30]: I need to look into a less lazy name here

**Review by gkartalis — COMMENTED** [review]
**question:** since we do have the Screen and Screen.AnimatedHeader components in palette are we thinking of embedding this logic there as well?

**Comment by ArtsyOpenSource** [comment]
### This PR contains the following changes:

- Cross-platform user-facing changes (add fancy collapsible header logic - mounir)
<p align="right">
  Generated by :no_entry_sign: <a href="https://danger.systems/js">dangerJS</a> against 410c2cd63be8884412ba78997628d4c230031229
</p>

**Review by olerichter00 — COMMENTED** [review]
This looks great! 🌟 

Doesn't [PageWithSimpleHeader](https://github.com/artsy/eigen/blob/0d20151ef900df94801f37a79b091c8794059c12/src/app/Components/PageWithSimpleHeader.tsx) has the same functionality (just without this great animation)?

**Comment by MounirDhahri** [comment]
> since we do have the Screen and Screen.AnimatedHeader components in palette are we thinking of embedding this logic there as well?

Nice idea, I think it should be possible to add this there as well. I had this here initially because it's only used in the `Activity` screen but I will talk to design to see if this is something we would like to add everywhere to include it in palette.

**Comment by MounirDhahri** [comment]
> This looks great! 🌟

Thank you!

> Doesn't [PageWithSimpleHeader](https://github.com/artsy/eigen/blob/0d20151ef900df94801f37a79b091c8794059c12/src/app/Components/PageWithSimpleHeader.tsx) has the same functionality (just without this great animation)?

Thanks for the feedback as well. Yes, this is accurate, but it's unfortunately not as nice as a UX as I believe this is. My goal here is to slightly elevate the UX and hopefully we can have this in many other screens once Barney approves it

**Comment by olerichter00** [comment]
> Thanks for the feedback as well. Yes, this is accurate, but it's unfortunately not as nice as a UX as I believe this is. My goal here is to slightly elevate the UX and hopefully we can have this in many other screens once Barney approves it.

Yes, your header component looks way better 🤩. Then, I will approve this. Feel free to add it to the new Activity screen. 

But anyway, I think we **should** update [PageWithSimpleHeader](https://github.com/artsy/eigen/blob/0d20151ef900df94801f37a79b091c8794059c12/src/app/Components/PageWithSimpleHeader.tsx) and the Palette's `Screen` component to use this new animation/implementation. If we only add it to the Activity screen, we will have an inconsistent header throughout the app, which isn't ideal.

**Comment by MounirDhahri** [comment]
I am on it 🤝

**Comment by MounirDhahri** [comment]
I will close this in favor of this one https://github.com/artsy/palette-mobile/pull/176

**Pull request closed without merging**